File: src/content/docs/topics/cs/coding-problems/stack/020-valid-parentheses-approach3.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function isValid(s: string): boolean {
    const pairs: Record<string, string> = { ')': '(', ']': '[', '}': '{' };
    const stack: string[] = [];
    for (const ch of s) {
        if (Object.values(pairs).includes(ch)) {
            stack.push(ch);
        } else {
            if (!stack.length || stack.pop() !== pairs[ch]) return false;
        }
    }
    return stack.length === 0;
}

assert(isValid('()') === true);
assert(isValid('()[]{}') === true);
assert(isValid('(]') === false);
assert(isValid('([)]') === false);
assert(isValid('{[]}') === true);
assert(isValid('') === true);
assert(isValid('(') === false);
assert(isValid(')') === false);
console.log('all tests pass');

File: src/content/docs/topics/cs/coding-problems/stack/020-valid-parentheses-approach3.py
def is_valid(s: str) -> bool:
    pairs = {")": "(", "]": "[", "}": "{"}
    stack = []
    for ch in s:
        if ch in pairs.values():
            stack.append(ch)
        else:
            if not stack or stack.pop() != pairs[ch]:
                return False
    return not stack

assert is_valid("()") == True
assert is_valid("()[]{}") == True
assert is_valid("(]") == False
assert is_valid("([)]") == False
assert is_valid("{[]}") == True
assert is_valid("") == True
assert is_valid("(") == False
assert is_valid(")") == False
print("all tests pass")
